### ISSUE TITLE
should expound work anywhere s/explain works?  - clojurescript error

### ISSUE THREAD

**Issue by joefromct** [bhb/expound#198]
Hi, 

Maybe i'm misunderstanding the scope of expound, however is it safe to assume that any data that works with `explain` should work with `expound`? 

For instance, does there appear to be something wrong with this?

``` clojure
user> (expound/expound ::app-db @app-db)   
;; => #object[TypeError TypeError: s.replace is not a function]

user> (s/explain ::app-db @app-db) 
{:id "0878b3d4", :content-type :org, :content "text "} - failed: (contains? % :name) in: [:notes 0] at: [:notes :cljs.spec.alpha/pred] spec: :zkcljs.db/note
{:id "0878b3d4", :content-type :org, :content "A text node here blah blah"} - failed: (contains? % :name) in: [:notes 1] at: [:notes :cljs.spec.alpha/pred] spec: :zkcljs.db/note
{:from-id "0878b3d4", :to "0878b3d4"} - failed: (contains? % :to-id) in: [:relations 0] at: [:relations :cljs.spec.alpha/pred] spec: :zkcljs.db/relation
[{:id "0878b3d4", :content-type :org, :content "text "} {:id "0878b3d4", :content-type :org, :content "A text node here blah blah"}] - failed: nil? in: [:notes] at: [:notes :cljs.spec.alpha/nil] spec: :zkcljs.db/notes
[{:from-id "0878b3d4", :to "0878b3d4"}] - failed: nil? in: [:relations] at: [:relations :cljs.spec.alpha/nil] spec: :zkcljs.db/relations
;; => nil  

```

This is just for a toy clojurescript re-frame app, but i'm surprised i'm seeing this.

My shadow-cljs.edn has `[expound "0.8.5"]` as a dependency.

**Comment by bhb** [comment]
@joefromct That's correct: you should be able to replace `explain` with `expound` and it should just work. I would expect your example to work without error.

I'm not sure if this is a bug in Expound that only occurs with your specific combination of spec and data, or perhaps some incompatibility with Expound and Shadow that I'm unaware of.

Does this always occur, or only with certain specs and data structures? e.g. does `(expound/expound string? {})` work?

Does the same issue occur for your specs and data if you try it outside Shadow. For instance, does it repro in the following REPL?

`clj -Sdeps '{:deps {org.clojure/clojurescript {:mvn/version "1.10.764"} expound {:mvn/version "0.8.5"}}}' -m cljs.main -re node`

**Comment by joefromct** [comment]
Hi, Apologies for not getting back to you sooner.  Thanks for the reply.

This seems to work as expected:

This seems to work as expected: 
``` clojure
(expound/expound string? {}) 


-- Spec failed --------------------

  {}

should satisfy

  string?

-------------------------
Detected 1 error
```

However, when running over my app-db in re-frame it must not likes something in there; as stated, s/explain seems to work: 

```
(expound/expound ::app-db @app-db) 

;; => #object[TypeError TypeError: s.replace is not a function]
 
```

**Comment by joefromct** [comment]
Here's some additional details on part of my spec and data that strangely yields the error:  


``` clojure 
  ;; some test data:
  (def test-data
    {:tags nil,
     :notes
     [{:id "0878b3d4", :content-type :org, :content "text "}
      {:id "0878b3d4", :content-type :org, :content "A text node here blah blah"}],
     :relations [{:from-id "0878b3d4", :to "0878b3d4"}]})

  ;; this looks ok:
  (s/explain-str ::app-db test-data)

  "{:id \"0878b3d4\", :content-type :org, :content \"text \"} - failed: (contains? % :name) in: [:notes 0] at: [:notes :cljs.spec.alpha/pred] spec: :zkcljs.db/note\n{:id \"0878b3d4\", :content-type :org, :content \"A text node here blah blah\"} - failed: (contains? % :name) in: [:notes 1] at: [:notes :cljs.spec.alpha/pred] spec: :zkcljs.db/note\n{:from-id \"0878b3d4\", :to \"0878b3d4\"} - failed: (contains? % :to-id) in: [:relations 0] at: [:relations :cljs.spec.alpha/pred] spec: :zkcljs.db/relation\n[{:id \"0878b3d4\", :content-type :org, :content \"text \"} {:id \"0878b3d4\", :content-type :org, :content \"A text node here blah blah\"}] - failed: nil? in: [:notes] at: [:notes :cljs.spec.alpha/nil] spec: :zkcljs.db/notes\n[{:from-id \"0878b3d4\", :to \"0878b3d4\"}] - failed: nil? in: [:relations] at: [:relations :cljs.spec.alpha/nil] spec: :zkcljs.db/relations\n"

  ;; this gives me the strange error:
  ;;

  (expound/expound ::app-db test-data)
  #object[TypeError TypeError: s.replace is not a function]

```

Here's the relevant bits of the spec at the moment:
``` clojure
(s/def ::id (s/or :string string?
                :uuid uuid?))

(s/def ::kind #{:spec :app})

(s/def ::error (s/keys :req-un [::id ::kind]
                       :opt-un [::content]))
(s/def ::errors (s/coll-of ::error))

(s/def ::focus-areas any?)

(s/def ::tag keyword?)
(s/def ::tags (s/nilable (s/coll-of ::tag)))
(s/def ::content-type #{:org :markdown})
(s/def ::content string?)

(s/def ::note (s/keys :req-un [::id
                               ::name
                               ::content-type
                               ::content]
                      :opt-un [::tags
                               ::deadline
                               ::scheduled]))

(s/def ::notes (s/nilable (s/coll-of ::note)))

(s/def ::from-id ::id)
(s/def ::to-id   ::id)
(s/def ::relation (s/keys :req-un
                          [::from-id
                           ::to-id]))
(s/def ::relations
  (s/nilable (s/coll-of ::relation)))

(s/def ::focus-area any?)

(s/def ::pending any?)

(s/def ::app-db (s/keys :req-un [::tags
                                 ::notes
                                 ::relations]
                        :opt-un [::errors
                                 ::pending]))
```

**Comment by joefromct** [comment]
I think i have a smaller reproducible error here; The core of the issues is that expound behaves differently when I accidentally leave out a key from my app-db: 

``` clojure
  (require '[expound.alpha :as expound]
           '[clojure.spec.alpha :as s])

  ;; some specs
  (s/def ::key-a string?)
  (s/def ::key-b string?)
  (s/def ::keys (s/keys :req-un [::key-a
                                 ::key-b]))

  ;; Works:
  (s/explain ::keys {:key-a "1" :key-b "2"})
  (expound/expound ::keys {:key-a "1" :key-b "2"})

  ;; Now, redefine ::keys to have a key I forgot to define on it's own as a
  ;; field:
  (s/def ::keys (s/keys :req-un [::key-a
                                 ::key-b
                                 ::i-do-not-exist]))

  (s/explain-str ::keys {:key-a "1" :key-b "2"})
;;=> "{:key-a \"1\", :key-b \"2\"} - failed: (contains? % :i-do-not-exist) spec:
;;:zkcljs.db/keys\n"


(expound/expound-str ::keys {:key-a "1" :key-b "2"}) 
;; => #object[TypeError TypeError: s.replace is not a function]

```

**Comment by bhb** [comment]
@joefromct Thanks for the repro! I can confirm this reproduces for me on CLJS.

**Comment by bhb** [comment]
@joefromct Thanks for reporting this. If it's convenient in your project setup to depend on a specific SHA, this should be fixed in 121006405379b472a78317413aaec43896c160d6. If not, it'll be included in the next release (after I fix at least one other issue).

**Comment by bhb** [comment]
Fixed in https://github.com/bhb/expound/pull/200

**Comment by bhb** [comment]
@joefromct Thanks for your help on this. I just released 0.8.6, which includes this fix.
